pysnmp/carrier/asyncore/stream/base.py

```diff
@@ -0,0 +1,187 @@
+#
+# This file is part of pysnmp software.
+#
+# Copyright (c) 2005-2018, Ilya Etingof <etingof@gmail.com>
+# License: http://snmplabs.com/pysnmp/license.html
+#
+# Author: Edgar Sousa <edg@edgsousa.xyz>
+#
+import errno
+import socket
+import struct
+import sys
+
+from pysnmp import debug
+from pysnmp.carrier import sockmsg, error
+from ..dgram.base import DgramSocketTransport
+
+# Ignore these socket errors
+sockErrors = {errno.ESHUTDOWN: True,
+              errno.ENOTCONN: True,
+              errno.ECONNRESET: False,
+              errno.ECONNREFUSED: False,
+              errno.EAGAIN: False,
+              errno.EWOULDBLOCK: False,
+              errno.WSAEWOULDBLOCK: False,}
+
+if hasattr(errno, 'EBADFD'):
+    # bad FD may happen upon FD closure on n-1 select() event
+    sockErrors[errno.EBADFD] = True
+
+
+class StreamSocketTransport(DgramSocketTransport):
+    sockType = socket.SOCK_STREAM
+
+    def __init__(self, sock=None, sockMap=None):
+        DgramSocketTransport.__init__(self, sock, sockMap)
+        self.__outQueue = []
+        self._sendto = lambda s, b, a: s.send(b)
+
+        def __recvfrom(s, sz):
+            header_data = bytearray(s.recv(2))
+            if header_data:
+                #1st byte is 0x30 - sequence
+                seq_len = header_data[1]
+                if seq_len & 0x80 != 0:  # BER long definite length
```

Comment:
Finding start of message in the stream is hard indeed...

It would be cleaner if we keep transport code independent from the data being transferred. That would involve a bit of refactoring in the protocol engine, though.

[Here](https://github.com/etingof/pysnmp/blob/acd16a65f89b0165b881c4f33ee5802607583c3f/pysnmp/entity/engine.py#L149) is where the read message goes from transport/networking code into the protocol engine. I am thinking if we make `receiveMessage` returning unused bytes (as returned by the underlying pyasn1 decoder) and "unread" them [back](https://github.com/etingof/pysnmp/blob/master/pysnmp/carrier/asyncore/dgram/base.py#L163) into some internal queue like [__inQueue](https://github.com/etingof/pysnmp/blob/master/pysnmp/carrier/asyncore/dgram/base.py#L34), but (in case of stream transport) it should probably just a (growing/shrinking) string rather than a list...? Just like you have it in `remaining_bytes`.

That would let us remove this header finding heuristics completely. And, most importantly, make searching way more reliable.

One of the interesting problems along this way would be to detect and eventually reset the growing `inQueue` if decoder repeatedly fails to decode it. May be this could be based on a reasonable SNMP message size, like if the buffer grows beyond 4K - drop it.

Does it make sense?